### PR TITLE
Package notifier to interact with gorush notifications

### PR DESCRIPTION
**Pull request by adriacidre**
First small iteration against moving push notifications to use [gorush](https://github.com/appleboy/gorush)

This pull request introduces a new package *(notifier)* so we have an interface to setup and send messages through gorush.

Important changes:
- [x] Introduced notifier package to manage push notifications

### CODE REVIEW

**Comment by dshulyak** [comment]
can we use http api and avoid adding this thingy to deps?

https://github.com/appleboy/gorush/blob/master/README.md#post-apipush

<blockquote><div> GitHub</div><div><strong><a href="https://github.com/appleboy/gorush">appleboy/gorush</a></strong></div><div>gorush - A push notification server written in Go (Golang).</div></blockquote>

**Review by dshulyak — CHANGES_REQUESTED** [review]
i will add :x: cause i don't see it really necessary to add gorush dep with all other 100k of deps for such simple request

**Comment by adriacidre** [comment]
yep @dshulyak totally makes sense, i'll try with a simple http call

File: notifier/notifier.go [hunk 1]
```diff
@@ -0,0 +1,48 @@
...
+import (
+	"context"
+	"log"
+
+	"github.com/appleboy/gorush/rpc/proto"
+
+	"google.golang.org/grpc"
+)
+
+const fcmPlatform = 2
```
Comment (dshulyak) [inline notifier/notifier.go]: shouldn't it be 1 for ios?

File: notifier/notifier.go [hunk 2]
```diff
@@ -0,0 +1,48 @@
+package notifier
+
+import (
+	"context"
+	"log"
+
+	"github.com/appleboy/gorush/rpc/proto"
```
Comment (adambabik) [inline notifier/notifier.go]: This package has just one Go file and none `gorush` dependencies. Can we include only it in `Gopkg.toml`?

Comment (dshulyak) [reply notifier/notifier.go]: isn't it already like this?

```
+  name = "github.com/appleboy/gorush"
+  packages = ["rpc/proto"]
```

i think that most of the deps are from grpc that is also imported one line below

but the same should b doable with regular http client (please see comment above)

**Comment by adriacidre** [comment]
Ok guys, not sure why we went for grpc as a first option. I switched to basic http client now without any single dependency :)

File: notifier/notifier.go [hunk 3]
```diff
@@ -0,0 +1,103 @@
...
+type Response struct {
+	Logs []struct {
+		Type  string `json:"type"`
+		Error string `json:"error"`
+	} `json:"logs"`
+}
+
+// Send a push notification to given devices.
+func (n *Notifier) Send(notifications []*Notification) error {
+	url := n.url + "/api/push"
```
Comment (dshulyak) [inline notifier/notifier.go]: move to constant?

File: notifier/notifier.go [hunk 4]
```diff
@@ -0,0 +1,103 @@
...
+		return err
+	}
+
+	res, err := n.doRequest(url, body)
+	if err != nil {
+		return err
+	}
+
+	if len(res.Logs) > 0 {
+		if res.Logs[0].Type == "failed-push" {
```
Comment (dshulyak) [inline notifier/notifier.go]: constant?

File: notifier/notifier.go [hunk 5]
```diff
@@ -0,0 +1,103 @@
...
+	body, err = ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return
+	}
+
+	return res, json.Unmarshal(body, &res)
+}
+
+// Close the current connection.
+func (n *Notifier) Close() error {
```
Comment (dshulyak) [inline notifier/notifier.go]: is it a part of the interface?

**Review by dshulyak — APPROVED** [review]
thanks 👍

File: notifier/notifier.go [hunk 6]
```diff
@@ -0,0 +1,103 @@
...
+	Android = 2
+)
+
+type httpClient interface {
+	Do(req *http.Request) (*http.Response, error)
+}
+
+// Notifier handles android and ios push notifications.
+type Notifier struct {
+	client httpClient
```
Comment (adambabik) [inline notifier/notifier.go]: Any reason for `httpClient` type instead of `*http.Client`?

File: notifier/notifier.go [hunk 7]
```diff
@@ -0,0 +1,103 @@
...
+	}
+
+	res, err := n.doRequest(url, body)
+	if err != nil {
+		return err
+	}
+
+	if len(res.Logs) > 0 {
+		if res.Logs[0].Type == "failed-push" {
+			return errors.New(res.Logs[0].Error)
```
Comment (adambabik) [inline notifier/notifier.go]: This may be cryptic. Maybe `fmt.Errorf("notifier error: %v", res.Logs[0].Error)`?

File: notifier/notifier.go [hunk 8]
```diff
@@ -0,0 +1,103 @@
...
+
+	return err
+}
+
+func (n *Notifier) doRequest(url string, body []byte) (res Response, err error) {
+	req, err := http.NewRequest("POST", url, bytes.NewBuffer(body))
+	if err != nil {
+		return
+	}
+	req.Header.Set("X-Custom-Header", "myvalue")
```
Comment (adambabik) [inline notifier/notifier.go]: Is it needed?

File: notifier/notifier.go (hunk 5 above)
Comment (adriacidre) [reply notifier/notifier.go]: nope, just a deprecated method, removing it now

File: notifier/notifier_test.go [hunk 9]
```diff
@@ -0,0 +1,58 @@
...
+	},
+	{
+		response: `{"counts":1,"success":"ok"}`,
+		err:      nil,
+	},
+}
+
+func (s *NotifierTestSuite) TestExample() {
+	for _, tt := range flagtests {
+		var apiStub = httptest.NewServer(http.HandlerFunc(func(w http.ResponseWriter, r *http.Request) {
```
Comment (adambabik) [inline notifier/notifier_test.go]: You may want to call `defer apiStub.Close()`.

File: notifier/notifier_test.go [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+		response: `{"counts":1,"logs":[{"type":"failed-push","platform":"android","token":"1111111","message":"Hello world","error":"invalid registration token"}],"success":"ok"}`,
+		err:      errors.New("invalid registration token"),
+	},
+	{
+		response: `{"counts":1,"success":"ok"}`,
+		err:      nil,
+	},
+}
+
+func (s *NotifierTestSuite) TestExample() {
```
Comment (adambabik) [inline notifier/notifier_test.go]: Maybe we can give it a more informative name? :P

**Review by adambabik — APPROVED** [review]

File: notifier/notifier.go (hunk 6 above)
Comment (adriacidre) [reply notifier/notifier.go]: [httpClient](https://github.com/status-im/status-go/pull/910/files#diff-4063493f3b3cf719235a092fd180dd56R20) is an interface :)

File: notifier/notifier.go (hunk 8 above)
Comment (adriacidre) [reply notifier/notifier.go]: nope, sorry 😂 
All changes implemented, cheers mate.

File: notifier/notifier.go (hunk 6 above)
Comment (adambabik) [reply notifier/notifier.go]: Right, but I just wonder what was your motivation to create this interface and use it instead of providing a concrete type.

Comment (adriacidre) [reply notifier/notifier.go]: ah, sorry, again that's missing from a previous approach 😓 
fixed now

File: notifier/notifier.go (hunk 1 above)
Comment (pedropombeiro) [reply notifier/notifier.go]: As I understand it (maybe I'm wrong), the 2 identifies FCM and 1 identifies APNS, and we should be able to continue targeting both Android and iOS with FCM (2). We'll need to eventually confirm this though.

Comment (adriacidre) [reply notifier/notifier.go]: @PombeirP yep i had the same feeling, but [documentation](https://github.com/appleboy/gorush#post-apipush) seems to be clear about this:
```
Simple send iOS notification example, the platform value is 1
Simple send Android notification example, the platform value is 2
```
anyway, `Send` method will accept now the platform value, so we can set platform as default or not from the package consumer.

Comment (pedropombeiro) [reply notifier/notifier.go]: @adriacidre yeah, sometimes documentation is simplistic though. Unfortunately, I don't have an iOS device to test this with a POST request.

Comment (adriacidre) [reply notifier/notifier.go]: @PombeirP I have an IPAD, so i should be able test it

File: notifier/notifier.go [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+	"bytes"
+	"encoding/json"
+	"errors"
+	"io/ioutil"
+	"log"
+	"net/http"
+)
+
+const (
+	// IOS identifier for an ios mnotification.
```
Comment (pedropombeiro) [inline notifier/notifier.go]: nits: typo in notification, casing in iOS

**Review by pedropombeiro — APPROVED** [review]

**Pull request merged**